source/auth/auth.rst

```diff
@@ -683,6 +683,176 @@ mechanism
 mechanism_properties
 	MUST NOT be specified.
 
+MONGODB-IAM
+~~~~~~~~~~~
+
+:since: 4.4
+
+MONGODB-IAM authenticates using AWS IAM credentials or temporary IAM credentials 
+assigned to an EC2 machine or ECS task.
+
+MONGODB-IAM requires that a client create a randomly generated nonce. It is 
+imperative, for security sake, that this be as secure and truly random as possible. 
+
+All messages between MongoDB clients and servers are sent as base64 encoded BSON V1.1 Objects. 
+All fields in these messages have a "short name" which is used in the serialized 
+BSON representation and a human readable "friendly name" which is used in this document. They are as follows:
+
+==== ==================== ========== ============================================================================================================================================== 
+Name Friendly Name        Type       Description
+==== ==================== ========== ==============================================================================================================================================
+r    client nonce         BinData    32 byte cryptographically secure random number 
+p    gs2-cb-flag          32-bit int The integer representation of the ASCII charater 'n' or 'y', i.e., ``110`` or ``121``
+s    server nonce         BinData    64 bytes total. The first 32 bytes from the client nonce and the second 32 bytes from the server nonce
+h    sts hosts            string     DNS Host name of the STS service 
+a    auth_header          string     Authorization header for `AWS Signature Version 4 <https://docs.aws.amazon.com/general/latest/gr/signature-version-4.html?shortFooter=true>`_
+d    X-AMZ-Date           string     Current date in UTC -see `AWS Signature Version 4 <https://docs.aws.amazon.com/general/latest/gr/signature-version-4.html?shortFooter=true>`_
+t    X-AMZ-Security-Token string     Optional AWS security token
+==== ==================== ========== ============================================================================================================================================== 
+
+Drivers MUST NOT advertise support for channel binding, as the server does
+not support it and legacy servers may fail authentication if drivers advertise
+support. The client-first-message MUST set ``p`` to the integer representation 
+of the ASCII character ``n``, i.e., ``110``.
+
+Conversation
```

Comment:
So this protocol is not like SCRAM. There is no magic letters or equal signs, it is BSON encoded as base64. My spec was not clear that the examples where JSON representations. I included the JSON versions since it was more readable. I have updated my spec to be just base64.

The actual exchange looks like this:
************************************************
Normative example of bytes sent during a successful authentication using regular AWS credentials. The lines that start with // are comments to indicate which message each string corresponds with.
// Client First:
NAAAAAVyACAAAAAAdzw1U2IwSEtgaWI0IUxZMVJqc2xuQzNCcUxBc05wZjIQcABuAAAAAA==
// Server First:
ZgAAAAVzAEAAAAAAdzw1U2IwSEtgaWI0IUxZMVJqc2xuQzNCcUxBc05wZjIGS0J9EgLwzEZ9dIzr/hnnK2mgd4D7F52t8g9yTC5cIAJoABIAAABzdHMuYW1hem9uYXdzLmNvbQAA
// Client Second:
LQEAAAJhAAkBAABBV1M0LUhNQUMtU0hBMjU2IENyZWRlbnRpYWw9QUtJQUlDR1ZMS09LWlZZM1gzREEvMjAxOTExMDcvdXMtZWFzdC0xL3N0cy9hd3M0X3JlcXVlc3QsIFNpZ25lZEhlYWRlcnM9Y29udGVudC1sZW5ndGg7Y29udGVudC10eXBlO2hvc3Q7eC1hbXotZGF0ZTt4LW1vbmdvZGItZ3MyLWNiLWZsYWc7eC1tb25nb2RiLXNlcnZlci1ub25jZSwgU2lnbmF0dXJlPWFiNjJjZTFjNzVmMTljNGM4YjkxOGIyZWQ2M2I0NjUxMjc2NWVkOWI4YmI1ZDc5YjM3NGFlODNlZWFjMTFmNTUAAmQAEQAAADIwMTkxMTA3VDAwMjYwN1oAAA==
************************************************

You notice there is no r=, s=, h=, etc. It is just ugly raw base64 encoding of BSON.

************************************************
Non-normative example of above conversation converted from base64 to BSON and then to JSON (i.e. each message was processed by the pseudo code bsonToJson(decodeBase64(message)) )
// Client First:
{ 
     "r" : new BinData(0, "dzw1U2IwSEtgaWI0IUxZMVJqc2xuQzNCcUxBc05wZjI="),
     "p" : 110
}
// Server First:
{ 
    "s" : new BinData(0, "dzw1U2IwSEtgaWI0IUxZMVJqc2xuQzNCcUxBc05wZjIGS0J9EgLwzEZ9dIzr/hnnK2mgd4D7F52t8g9yTC5cIA=="),
    "h" : "sts.amazonaws.com"
}
// Client Second:
{ 
    "a" : "AWS4-HMAC-SHA256 Credential=AKIAICGVLKOKZVY3X3DA/20191107/us-east-1/sts/aws4_request, SignedHeaders=content-length;content-type;host;x-amz-date;x-mongodb-gs2-cb-flag;x-mongodb-server-nonce, Signature=ab62ce1c75f19c4c8b918b2ed63b46512765ed9b8bb5d79b374ae83eeac11f55",
    "d" : "20191107T002607Z"
}
************************************************

Reply:
Am I correct in understanding that the protocol exchange data starts out as a hash, this hash is serialized into bson, bson is base64 encoded, the result is placed into a server command which is another hash, that hash is bson encoded and the result is sent over the wire?

What is the rationale for this approach over placing auth data into the command hash as most/all of the existing authentication methods do? It seems to me that having auth data in the command hash would be significantly easier to debug and work with in general.

Reply:
You are partially correct. Your first part of your sentence is correct: "The protocol exchange starts out as a hash, this hash is serialized into bson, bson is base64 encoded". At this point, the client creates a new bson document and has a choice of storing the base64 encoded bson in the "payload" field as bindata or as a string (in the string case, it is again converted to base64 again). I have realized my first base64 encode is redundant in this case and I plan to remove it from the protocol.

``
// My first version
{ 
    "saslStart" : 1, 
    "mechanism" : "MONGODB-IAM", 
    "payload" : new BinData(0, "TkFBQUFBVnlBQ0FBQUFBQUpTb2pm..."),
}
where payload = base64encode(BSON({  { "r" : new BinData(0, "JSojfm1PJyQ7R34lTiV+V19rYD1JODA8ZT4ySzczMHQ="), "p" : 110 } }))
``

When you say: "over placing auth data into the command hash as most/all of the existing authentication methods do", what are you proposing as an alternative?

Something like this where the bson is nested as bson?
``
{ 
    "saslStart" : 1, 
    "mechanism" : "MONGODB-IAM", 
    "payload" : { "r" : new BinData(0, "JSojfm1PJyQ7R34lTiV+V19rYD1JODA8ZT4ySzczMHQ="), "p" : 110 }
}``
This would not work since all implementations require payload to be either bindata or string.

Or where payload is a string serialized in a different format like SCRAM's comma delimited format?
``
{ 
    "saslStart" : 1, 
    "mechanism" : "MONGODB-IAM", 
    "payload" : { "r=JSojfm1PJyQ7R34lTiV+V19rYD1JODA8ZT4ySzczMHQ=,p=n" }
}``

If I modeled if after SCRAM, I would need to worry about escaping commas. Also, in the server, we can easily parse BSON via IDL generated BSON parsers so it was a more natural fit.

Reply:
Done

Reply:
Thanks - Please resolve because I cannot